src/transport/raw/TCP.cpp

```diff
@@ -290,101 +268,82 @@ CHIP_ERROR TCPBase::SendAfterConnect(const PeerAddress & addr, System::PacketBuf
-            // on purpose:
-            //   - we already consumed the packet size above
-            //   - there is no reason to believe that an error would not occur again on the
-            //     same parameters (errors are likely not transient)
-            //   - this guarantees data is received and progress is made.
-            err = ProcessSingleMessageFromBufferHead(peerAddress, buffer, messageSize);
-            buffer->ConsumeHead(messageSize);
+            // The buffer contains a complete message.
+            consumed += messageSize + kPacketSizeBytes;
+            buffer.Consume(kPacketSizeBytes);
+            err = ProcessSingleMessage(peerAddress, buffer, messageSize);
             SuccessOrExit(err);
-
-            err = endPoint->AckReceive(messageSize);
-            SuccessOrExit(err);
-            continue;
-        }
-
-        // Buffer is incomplete if we reach this point
-        if (buffer->HasChainedBuffer())
-        {
-            buffer->CompactHead();
-            continue;
         }
-
-        if (messageSize > 0)
-        {
-            // Open the receive window just enough to allow the remainder of the message to be received.
-            // This is necessary in the case where the message size exceeds the TCP window size to ensure
-            // the peer has enough window to send us the entire message.
-            uint16_t neededLen = static_cast<uint16_t>(messageSize - buffer->DataLength());
-            err                = endPoint->AckReceive(neededLen);
-            SuccessOrExit(err);
-        }
-
-        // Buffer is incomplete and we cannot get more data
-        break;
     }
 
 exit:
+    // Ack the data consumed.
+    endPoint->AckReceive(consumed);
+
+    // Return any partial message to the endpoint, and wait for more.
     if (!buffer.IsNull())
     {
-        // Incomplete processing will be retried
         endPoint->PutBackReceivedData(std::move(buffer));
     }
 
     return err;
 }
 
+CHIP_ERROR TCPBase::ProcessSingleMessage(const PeerAddress & peerAddress, System::PacketBufferHandle & buffer, uint16_t messageSize)
+{
+    // We enter with `buffer` containing at least one full message, perhaps in a chain.
+    // `buffer->Start()` currently points to the message data.
+    // On exit, `buffer` will have had `messageSize` bytes consumed, no matter what.
+    System::PacketBufferHandle message;
+    if (buffer->DataLength() == messageSize)
+    {
+        // In this case, the head packet buffer contains exactly the message.
+        // This is common because typical messages fit in a network packet, and are delivered as such.
+        // Peel off the head to pass upstream.
+        message = buffer.Retain();
+        buffer.Advance();
+    }
+    else
+    {
+        // The message is either longer or shorter than the head buffer.
+        // In either case, copy the message to a fresh linear buffer to pass upstream.
```

Comment:
So in the "shorter" case we are avoiding extra code to cope with that case, and mutation of the buffer length and whatnot, at the cost of doing a copy we don't really need to do?  I guess we also avoid the "pass consumer a buffer it doesn't really own, and might try to reuse to send a response" problem?  It might be worth adding some documentation for why copying in the "shorter" case is the right tradeoff.

Reply:
OK. Copying in the shorter case (which would occur, I guess, if the other end sends multiple CHIP messages in one TCP message) is meant to guard against the upper layer doing something destructive with the shared buffer chain (`TCPEndPoint.mRcvQueue`) or its contents beyond the current message.

Reply:
In particular this solves #4869 _Need to sort out what happens with TCP if the upper layers decide to reuse the buffer they receive to do a send_

Reply:
Right. I'm just saying we should document this clearly here, so someone doesn't decide to "optimize" this in the future.

Reply:
Sorry, thought I'd changed the comment already.